api/sirsidynix_authentication_provider.py

```diff
@@ -0,0 +1,332 @@
+                object_session(library), self.Keys.LIBRARY_ID, library, integration
+            ).value
+        )
+        self.sirsi_library_prefix = (
+            ConfigurationSetting.for_library_and_externalintegration(
+                object_session(library), self.Keys.LIBRARY_PREFIX, library, integration
+            ).value
+        )
+
+    def remote_authenticate(self, username: str, password: str) -> PatronData | bool:
+        """Authenticate this user with the remote server."""
+        data = self.api_patron_login(username, password)
+        if not data:
+            return False
+
+        return SirsiDynixPatronData(
+            username=username,
+            authorization_identifier=username,
+            permanent_id=data.get("patronKey"),
+            session_token=data.get("sessionToken"),
+            complete=False,
+        )
+
+    def remote_patron_lookup(
+        self, patron_or_patrondata: Patron | SirsiDynixPatronData
+    ) -> None | SirsiDynixPatronData:
+        """Do a remote patron lookup, this method can only lookup a patron with a patrondata object
+        with a session_token already setup within it.
+        This method also checks all the reasons that a patron may be blocked for.
+        """
+        # We cannot do a remote lookup without a session token
+        if not isinstance(patron_or_patrondata, SirsiDynixPatronData):
+            return None
+        elif not patron_or_patrondata.session_token:
+            return None
+
+        patrondata = patron_or_patrondata
+        # Pull and parse the basic patron information
+        data = self.api_read_patron_data(
+            patron_key=patrondata.permanent_id,
+            session_token=patrondata.session_token,
+        )
+        if not data or "fields" not in data:
+            return None
+
+        fields: dict = data["fields"]
+        patrondata.personal_name = fields.get("displayName")
+        patron_type: str = fields["patronType"].get("key", "")
+
+        patrondata.external_type = patron_type
+
+        # Basic block reasons
+
+        # Does the patron type start with the library prefix
+        # This ensures the patron is a member of the library
+        # they are interacting with
+        if not patron_type.startswith(self.sirsi_library_prefix):
+            patrondata.block_reason = SirsiBlockReasons.INCORRECT_LOCATION
+            return patrondata
+
+        if not fields.get("approved", False):
+            patrondata.block_reason = SirsiBlockReasons.NOT_APPROVED
+            return patrondata
+
+        # Remove the library prefix from the patron type
+        # we are left with the patron "suffix"
+        # This suffix can be part of the blocked patron types list
+        patron_suffix = patron_type[len(self.sirsi_library_prefix) :]
+        if patron_suffix in self.sirsi_disallowed_prefixes:
+            patrondata.block_reason = SirsiBlockReasons.PATRON_BLOCKED
+            return patrondata
```

Comment:
I'd rather see this enforced the `enforce_library_identifier_restriction` function https://github.com/ThePalaceProject/circulation/blob/03fe8e5d02bce41d3d68147d29ff0d8bd30b61af/api/authenticator.py#L1546-L1584

This is how we are enforcing patron restrictions in other integration.

Reply:
The SD restrictions are a little different than what `_restriction_matches` allows (one positive, one negative), so I've implemented it differently for SD.